staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -601,6 +672,46 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+// getOrCreateServingInfoForWrite is getOrCreateServingInfoFor, but blocks CR
+// write requests until the storage version update is processed, aborted, or
+// timed-out.
+// If the storage version update is processed, unblock and allow the write request.
+// If the storage version update is aborted, unblock and reject the write request.
+// If the storage version update timed-out, unblock and allow the write request.
+func (r *crdHandler) getOrCreateServingInfoForWrite(uid types.UID, name string) (*crdInfo, error) {
+	ret, err := r.getOrCreateServingInfoFor(uid, name)
+
+	// Double check the feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) &&
+		ret.storageVersionUpdateProcessedCh != nil {
+		// NOTE: currently the graceful CRD deletion waits 1s for in-flight requests
+		// to register themselves to the wait group. Ideally the storage version update should
+		// not cause the requests to miss the 1s window; otherwise the requests may
+		// fail ungracefully (e.g. it may happen if the CRD was deleted immediately after the
+		// first CR request establishes the underlying storage).
+		select {
+		case <-ret.storageVersionUpdateProcessedCh:
+		case <-ret.storageVersionUpdateAbortedCh:
+			// Storage version update failed because the CRD was deleted. Here
+			// we return a 404 to fail this in-flight request.
+			return nil, apierrors.NewNotFound(schema.GroupResource{
+				Group:    ret.spec.Group,
+				Resource: ret.spec.Names.Plural,
+			}, name)
+		// Unblock the requests if the storage version update takes a long time, otherwise
+		// CR requests may stack up and overwhelm the API server.
+		// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+		case <-time.After(ret.storageVersionUpdateTimeout.Sub(time.Now())):
```

Comment:
time.After hopefully works with negative durations 🤞

Reply:
yes, zero and negative durations are ignored